lib/bolt/logger.rb

```diff
@@ -123,18 +130,54 @@ def self.reset_logging
       Logging.reset
     end
 
-    def self.warn_once(type, msg)
+    # Issues a warning unless the user has suppressed it.
+    def self.warn(id:, msg:, src: self)
+      return if @disable_warnings.include?(id)
+
       @mutex.synchronize do
-        @logger ||= Bolt::Logger.logger(self)
-        if @warnings.add?(type)
-          @logger.warn(msg)
+        Bolt::Logger.logger(src).warn("#{msg} [ID: #{id}]")
+      end
+    end
+
+    # Issues a warning unless the user has suppressed it,
+    # or a warning with the same ID has already been issued.
+    def self.warn_once(id:, msg:, src: self)
+      return if @disable_warnings.include?(id)
+
+      @mutex.synchronize do
+        if @warnings.add?(id)
+          Bolt::Logger.logger(src).warn("#{msg} [ID: #{id}]")
         end
       end
     end
 
-    def self.deprecation_warning(type, msg)
-      @analytics&.event('Warn', 'deprecation', label: type)
-      warn_once(type, msg)
+    # Issues a deprecation warning unless the user has suppressed
+    # it. Submits analytics for the deprecation regardless of
+    # whether the warning is suppressed.
+    def self.deprecation_warning(id:, msg:, src: self)
+      @analytics&.event('Warn', 'deprecation', label: id)
+      warn(id: id, msg: msg, src: src)
```

Comment:
I believe this was originally `warn_once` to handle deprecations logged when connecting to targets (specifically PowerShell 2 deprecations). I switched this over to just use `warn` to handle cases where a user has a deprecated config option in multiple config layers.

Leaving this as `warn_once` for the config option case would result in the warning for each config layer having its own ID (`apply_settings_user`, `apply_settings_system`, `apply_settings_default`), otherwise users wouldn't see a warning for each config layer. However, using `warn` results in _every_ target with PowerShell 2 logging a deprecation warning... I wonder if we might want a separate `deprecation_warn_once` method?

Reply:
`deprecation_warn_once` is probably the simplest way to solve this. I think we may eventually want to have a warning hierarchy of sorts, where one ID can refer to multiple warnings (so `apply_settings` would map to `apply_settings_user`, `apply_settings_system`, and `apply_settings_default`, but users could also disable any of those warnings individually if they wanted to suppress `apply_settings_default` but not `apply_settings_user`) but adding that now really complicates things and I'm not sure it's needed yet.